nativelib/src/main/resources/scala-native/gc/immix_commix/GCRoots.c

```diff
@@ -0,0 +1,42 @@
+#include "GCRoots.h"
+
+#include <stdio.h>
+
+void GC_Roots_Add(GC_Roots **head, AddressRange range) {
+    // Prepend the node with given range to the head of linked list of GC roots
+    GC_Roots *node = (GC_Roots *)malloc(sizeof(GC_Roots));
+    node->range = range;
+    node->next = NULL;
+    GC_Roots *currentHead = *head;
+
+    do {
+        node->next = currentHead;
+    } while (!atomic_compare_exchange_strong((_Atomic(GC_Roots *) *)head,
+                                             &currentHead, node));
+}
+
+void GC_Roots_RemoveByRange(GC_Roots **head, AddressRange range) {
+    GC_Roots *current = *head;
+    GC_Roots *prev = NULL;
+    while (current != NULL) {
+        if (AddressRange_Contains(range, current->range)) {
+            bool detached = false;
+            if (prev == NULL) {
+                *head = current->next;
+                GC_Roots *expected = NULL;
+                detached = atomic_compare_exchange_strong(
+                    (_Atomic(GC_Roots *) *)head, &expected, current->next);
+            } else {
+                prev->next = current->next;
```

Comment:
There is a risk that if two threads try to concurrently remove items from the singly linked list, one of them will access freed memory on this line.

Suppose the singly linked list has the following nodes: head -> A -> B -> C -> D;

Thread 1 wants to remove C.  It calls `GC_Roots_RemoveByRange`. `current` is now pointing to C and `prev` is pointing to B. Thread 1 is just about to execute line 30, but context switching happened.

Then thread 2 starts executing. It wants to remove B.  It called `GC_Roots_RemoveByRange` and it quickly finished.  Now the linked list looks like: head -> A -> C -> D, while B is freed.

Then thread 1 wakes up, and continues to execute line 30.  Now `prev` already points to freed memory.  The atomic compare-exchange doesn't help because no matter whether it is successful or not, some freed memory have already been overwritten.

I guess the "custom GC roots" feature is for internal use for the ScalaNative runtime.  In this case, I think it will not be called very often.  It should be OK to use mutex for synchronisation.

Reply:
Thank you for this feedback and for bringing awareness. I agree that it might be a real issue and it would require improvements. The main purpose of custom GC roots was CaptureCalculus SafeZones experiments added in https://github.com/scala-native/scala-native/pull/3120. In the current implementation, it would add/remove gc roots once per usage of the SafeZone block in the code. 

I'll convert this comment into the dedicated issue